hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/BaseSparkCommitActionExecutor.java

```diff
@@ -418,4 +426,23 @@ public Partitioner getLayoutPartitioner(WorkloadProfile profile, String layoutPa
   protected void runPrecommitValidators(HoodieWriteMetadata<HoodieData<WriteStatus>> writeMetadata) {
     SparkValidatorUtils.runValidators(config, writeMetadata, context, table, instantTime);
   }
+
+  private int dynamicSampleRecordSize(JavaRDD<HoodieRecord<T>> inputRecords) {
+    int dynamicSampleRecordSize = config.getCopyOnWriteRecordSizeEstimate();
+    long inputRecordsCount = inputRecords.count();
+    if (inputRecordsCount == 0) {
+      LOG.warn("inputRecords is empty.");
+      return dynamicSampleRecordSize;
+    }
+    int maxSampleRecordNum = (int) Math.ceil(Math.min(inputRecordsCount * config.getRecordSizeDynamicSamplingRatio(), config.getRecordSizeDynamicSamplingMaxnum()));
+    try {
+      List<HoodieRecord<T>> sampleRecords = inputRecords.takeSample(false, maxSampleRecordNum);
+      dynamicSampleRecordSize = (int) (ObjectSizeCalculator.getObjectSize(sampleRecords) / maxSampleRecordNum);
```

Comment:
This is not the right way to estimate the object size -- this will provide us with the size of the Java object stored in memory (including pointers, etc).

Instead we should
 - Serialize all obtained records 
 - Divide total bytes by the number of records

Note, that some record impls like HoodieSparkRecord, HoodieAvroRecord already hold serialized representations so we don't need to serialize again in that case and can simply expose API (like `estimateSerializedSize`) in HoodieRecord interface to provide it

Reply:
@alexeykudinkin  thanks Alexey! Got it. I'll optimize this code